### PR TITLE
Documentation/policy: add label-based egress documentation

### PR DESCRIPTION
**Pull request by ianvernon**
Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #2556 

```release-note
Add label-based egress documentation
```

**How to test (optional)**:
Run `make render-docs` and go to `http://localhost:8080/policy/language.html` . The egress documentation will be there.

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: Documentation/policy/language.rst
Comment (manalibhutiyani) [inline Documentation/policy/language.rst]: nit: `toRequires`

**Review by aanm — CHANGES_REQUESTED** [review]
I think the documentation would look better if the policies existed in the repository as well

**Comment by ianvernon** [comment]
One day, I will add untracked files on the first try.

File: Documentation/policy/language.rst
Comment (ianvernon) [reply Documentation/policy/language.rst]: Fixed.

**Review by aanm — APPROVED** [review]

**Review by manalibhutiyani — APPROVED** [review]
Looks great!

File: examples/policies/l3/egress-allow-all/egress-allow-all.yaml
Comment (joestringer) [inline examples/policies/l3/egress-allow-all/egress-allow-all.yaml]: What's the thought behind calling it "from victim" here? I'd usually think of the victim as the one receiving unsolicited traffic, not the one sending it.

Comment (ianvernon) [reply examples/policies/l3/egress-allow-all/egress-allow-all.yaml]: I didn't really know what else to call it, I figured we'd keep it this way as a mirror image of ingress allow-all. I could also spin that it's also slightly uplifting to see that the victim can fight back from its attackers :P

**Comment by ianvernon** [comment]
test-me-please

File: examples/policies/l3/egress-allow-all/egress-allow-all.yaml
Comment (joestringer) [reply examples/policies/l3/egress-allow-all/egress-allow-all.yaml]: ;) While I appreciate the sentiment, I doubt that people looking at the policy with no context will understand it. It's not strongly important, but I suggest we either come up with witty names that make sense without context, or just name them generically.

Comment (ianvernon) [reply examples/policies/l3/egress-allow-all/egress-allow-all.yaml]: Changed it to frontend.

**Comment by ianvernon** [comment]
@joestringer  @cilium/docs please review.

File: Documentation/policy/language.rst
Comment (joestringer) [inline Documentation/policy/language.rst:42]: I wonder if the "send" / "receive" traffic descriptions here are a little too strict if read literally. It's a bit more abstract, the previous wording said "talk to" to imply that if it talks to the other endpoint, the other endpoint can talk back on the same channel, whereas "send traffic" implies unidirectional. But maybe I'm just reading too far into the wording.

Comment (ianvernon) [reply Documentation/policy/language.rst:42]: One counter I have to what you are saying is that we can still have label-based policy without connection tracking enabled; that is, we can disable connection tracking so that in order for an endpoint A to talk to another endpoint B, not only does A need to have an egress policy allowing it to talk to B, B needs a policy allowing ingress from A, and vice versa. I think not specifying whether communication is allowed in the reverse direction is good, because that is related to another configuration option (conntrack).

File: examples/policies/l3/egress-allow-all/egress-allow-all.json
Comment (raybejjani) [inline examples/policies/l3/egress-allow-all/egress-allow-all.json]: This doesn't match the yaml selector.

File: Documentation/policy/language.rst
Comment (raybejjani) [inline Documentation/policy/language.rst:125]: Would it be worthwhile to point out here that you can allow all egress but an endpoint with ingress rules would not allow anything to connect to it. This might be blatantly obvious (it is to us) but I don't have a sense for how much we want our doc sections to be independently complete.
A similar note might be useful in the examples above. In particular, noting that the egress allow policy allows `frontend` pod outbound connections to `backend` pods but the `backend` pod ingress policy must allow connections for stuff to work (the note can be much shorter, I'm just explaining).

**Review by raybejjani — APPROVED** [review]

File: Documentation/policy/language.rst
Comment (joestringer) [reply Documentation/policy/language.rst:42]: Fair counter, though let me counter counter ( ;D ) with this: What's the default? Shouldn't we document the behaviour as per the default operation of the daemon?

I'm not sure if the conntrack option you're referring to is even documented, and not only that, I honestly see it as a serious power user tweak because it breaks policies in a pretty fundamental way. You have to pretty much rewrite your entire policies to use it, and even then seems painful to me.

File: examples/policies/l3/egress-allow-all/egress-allow-all.json
Comment (joestringer) [reply examples/policies/l3/egress-allow-all/egress-allow-all.json]: Time to pull out https://github.com/cilium/1abel?

File: Documentation/policy/language.rst
Comment (ianvernon) [reply Documentation/policy/language.rst:42]: I disagree. Let's not make the documentation more complicated than it has to be. If we want to document the default behavior of the demon with connection tracking, let's get a separate section out to do so. This section specifically pertains to policy alone, not to daemon operations.

File: examples/policies/l3/egress-allow-all/egress-allow-all.json
Comment (ianvernon) [reply examples/policies/l3/egress-allow-all/egress-allow-all.json]: Will be fixed in latest upload.

File: Documentation/policy/language.rst
Comment (ianvernon) [reply Documentation/policy/language.rst:125]: I provided a general note for both ingress and egress, which I think is enough to point users in the right direction. Let me know what you think.

File: Documentation/policy/language.rst
Comment (joestringer) [reply Documentation/policy/language.rst:42]: Looking back over the sections as a whole, I agree it would make more sense to shift such detail out to a separate section.

**Comment by ianvernon** [comment]
test-docs-please

**Comment by ianvernon** [comment]
Merging, as I have received approval from all req'd codeowners and believe I have addressed all comments.

**Pull request merged**
